src/test/unit/io/json/json_ill_formed_test.cpp

```diff
@@ -0,0 +1,206 @@
+}
+
+TEST(errJson, jsonData_err_array_of_obj) {
+  std::string txt = "[ { \"foo\": 1}, { \"bar\": 1 } ]";
+  test_exception(txt, "Expecting JSON object, found array.");
+}
+
+TEST(errJson, jsonData_parse_mult_objects_err) {
+  std::string txt = "{ \"foo\": 1}{ \"bar\": 1 }";
+  test_exception(txt,
+                 "Error in JSON parsing \nat offset 11: \nThe document "
+                 "root must not be followed by other values.\n");
+}
+
+TEST(errJson, jsonData_empty_2D_array_1_0) {
+  std::string txt = "{ \"foo\" : [] }";
+  std::stringstream in(txt);
+  stan::json::json_data jdata(in);
+  std::vector<int> expected_vals_i;
+  std::vector<size_t> expected_dims;
+  expected_dims.push_back(1);
+  expected_dims.push_back(0);
+  test_empty_int_arr(jdata, "foo", expected_vals_i);
+  try {
+    jdata.validate_dims("test", "foo", "int", expected_dims);
+  } catch (const std::exception &e) {
+    FAIL();
+  }
+}
+
+TEST(errJson, jsonData_empty_3D_array_0_0_0) {
+  std::string txt = "{ \"foo\" : [] }";
+  std::stringstream in(txt);
+  stan::json::json_data jdata(in);
+  std::vector<int> expected_vals_i;
+  std::vector<size_t> expected_dims;
+  expected_dims.push_back(0);
+  expected_dims.push_back(0);
+  expected_dims.push_back(0);
+  test_empty_int_arr(jdata, "foo", expected_vals_i);
+  try {
+    jdata.validate_dims("test", "foo", "int", expected_dims);
+  } catch (const std::exception &e) {
+    FAIL();
+  }
+}
+
+TEST(errJson, jsonData_empty_3D_array_2_1_0) {
+  std::string txt = "{ \"foo\" : [] }";
+  std::stringstream in(txt);
+  stan::json::json_data jdata(in);
+  std::vector<int> expected_vals_i;
+  std::vector<size_t> expected_dims;
+  expected_dims.push_back(2);
+  expected_dims.push_back(1);
+  expected_dims.push_back(0);
+  test_empty_int_arr(jdata, "foo", expected_vals_i);
+  try {
+    jdata.validate_dims("test", "foo", "int", expected_dims);
+  } catch (const std::exception &e) {
+    FAIL();
+  }
+}
+
+TEST(errJson, jsonData_redefine_var_1) {
+  std::vector<std::string> json_path;
+  json_path = {
+      "src", "test", "unit", "io", "test_json_files", "redefine_vars_1.json"};
+  std::string filename = paths_to_fname(json_path);
+  std::string txt = file2str(filename);
+  test_exception(txt, "Attempt to redefine variable: x.");
+}
+
+TEST(errJson, jsonData_redefine_var_2) {
+  std::vector<std::string> json_path;
+  json_path = {
+      "src", "test", "unit", "io", "test_json_files", "redefine_vars_2.json"};
+  std::string filename = paths_to_fname(json_path);
+  std::string txt = file2str(filename);
+  test_exception(txt, "Attempt to redefine variable: x.");
+}
+
+TEST(errJson, jsonData_redefine_var_3) {
+  std::vector<std::string> json_path;
+  json_path = {
+      "src", "test", "unit", "io", "test_json_files", "redefine_vars_3.json"};
+  std::string filename = paths_to_fname(json_path);
+  std::string txt = file2str(filename);
+  test_exception(txt, "Attempt to redefine variable: x.");
+}
+
+TEST(errJson, jsonData_redefine_var_4) {
```

Comment:
This new file is great, thanks for adding. 

Could you also add some tests for:

- Non-regular arrays, e.g.  `"x": [ {"1": 0, "2": 1}, {"1": 1}]` and `"x": [ {"1": 0}, {"1": 1, "2": 1}]`
- Changing tuple types in an array, e.g. `"x": [ {"1": 0, "2": 1}, {"1": [1.5], "2":2}]`
- More redefines: What happens if I have a tuple `"x": {"1": [1,2,3]}` and then I add a JSON entry with the name `"x.1"`? Can you also add the reversed-order tests of the redefine_vars_1 and redefine_vars_4 tests (e.g. for vars_1, the order where the tuple is second, and for vars_4, the order where the array-of-tuples is first)

Reply:
In general I think we can't have too much testing of these edge cases, so I'm happy to generate more as well.

Reply:
> More redefines: What happens if I have a tuple "x": {"1": [1,2,3]} and then I add a JSON entry with the name "x.1"? What happens if the order is reversed and I have a json entry for "x.1" and then a tuple called "x" later?

this isn't possible - "x.1" is not a valid variable name - `{ "x.1" : 3.14 }` is invalid - or should be.

Reply:
It’s not a valid name in Stan, but someone could type it into a json file

Reply:
it looks like the JSON parser isn't checking for valid Stan variable name, so
```
{
    "x.1": 42,
    "x": {"1": 1, "2": 6.28}
}
```
parses just fine.
we could either disallow Stan variable names which aren't legal - the most robust thing to do would be to parse them, but then not save them.

Reply:
got a handy regex for legal Stan variable name in C++?

Reply:
I think `[a-zA-Z][a-zA-Z0-9_]*` should work. (Starts with a letter, can be alphanumeric plus underscores)

technically we also don’t allow it to end in an undescore but that is harder to check with just a regex

Reply:
I like the ignore option.
trying to work out how the event handler should do this gracefully.

Reply:
> Non-regular arrays, e.g. "x": [ {"1": 0, "2": 1}, {"1": 1}] and "x": [ {"1": 0}, {"1": 1, "2": 1}]
> Changing tuple types in an array, e.g. "x": [ {"1": 0, "2": 1}, {"1": [1.5], "2":2}]

These are tests by the unit tests `json_ill_formed.cpp` and the corresponding JSON files are in folder `src/test/unit/io/test_json_files/` names "inconsistent_array_tuples".